asim: add explanatory output to `assert` command

tbg opened the pull request:
I find that I have to check my understanding on this regularly,
so perhaps this output can help understand that `balance` checks
at each point in time that stores are close together, whereas `steady`
checks that for each store, its min/max does not deviate from its mean (over time),
in other words that the metric only changes within a small band. (Which does not
imply any kind of convergence across the cluster; each store might settle on its
own value for the metric).

Epic: CRDB-49117

tbg committed:
asim: add explanatory output to `assert` command

I find that I have to check my understanding on this regularly,
so perhaps this output can help understand that `balance` checks
at each point in time that stores are close together, whereas `steady`
checks that for each store, its min/max does not deviate from its mean (over time),
in other words that the metric only changes within a small band. (Which does not
imply any kind of convergence across the cluster; each store might settle on its
own value for the metric).

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/153585)
<!-- Reviewable:end -->

wenyihu6 reviewed (approved):
Thanks for doing this! A small nit but  <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153585#-:-O_Hux6fBNtxPottF-GG:b-2tpx35)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained

___
*[`pkg/kv/kvserver/asim/tests/testdata/non_rand/example_rebalancing.txt` line 35 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/153585#-O_HscGU8pkTZ3WJgOc8:-O_HscGV7ZDPYnggxgJ5:b-8jmidj) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2b3d8450c0dcde4ec296e710fc7b1e9027cdc0f/pkg/kv/kvserver/asim/tests/testdata/non_rand/example_rebalancing.txt#L35)):*
> ```help
> assertion stat=qps type=steady ticks=6 upper_bound=0.05
> ----
> asserting: qps stays < 1.05*mean_{ticks} at each store
> ```

nit: do you mind specifying that this assertion is also over the last 6 tickst, and that we are asserting qps is  `0.95*mean_{ticks} < qps < 1.05*mean_{ticks}` (alternatively, we can say `asserting: qps can only fluctuate within (±0.05) of its mean over the last 6 ticks`?)


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153585)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @wenyihu6)

___
*[`pkg/kv/kvserver/asim/tests/testdata/non_rand/example_rebalancing.txt` line 35 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/153585#-O_HscGU8pkTZ3WJgOc8:-O_IFkJU9cgi736zqGvN:b-sqbngy) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2b3d8450c0dcde4ec296e710fc7b1e9027cdc0f/pkg/kv/kvserver/asim/tests/testdata/non_rand/example_rebalancing.txt#L35)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

nit: do you mind specifying that this assertion is also over the last 6 tickst, and that we are asserting qps is  `0.95*mean_{ticks} < qps < 1.05*mean_{ticks}` (alternatively, we can say `asserting: qps can only fluctuate within (±0.05) of its mean over the last 6 ticks`?)
</blockquote></details>

Good callout, I massaged this a bit more and pushed a commit.

I need to figure out why this caused unrelated changes in the testdata output, which makes no sense to me.


<!-- Sent from Reviewable.io -->

tbg committed:
assertion: fix thresholds to allow endpoints

This is actually what they implement. For example, some
testdata assertions we have (and which pass) today verify
that a metric does not change at all by providing an upper_bound
of zero for a `steady` assertion, which can only pass if the
metric is constant (equal to the mean).

tbg committed:
asim: use exact_bound in a few cases

It's more intuitive in cases where we assert that the function is constant.

tbg committed:
assertion: address NaN issue

When a timeseries is constant zero, the computations were previously resulting
in NaNs, which we don't want here.

wenyihu6 commented:
*[`pkg/kv/kvserver/asim/tests/testdata/non_rand/example_rebalancing.txt` line 35 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/153585#-O_HscGU8pkTZ3WJgOc8:-O_MbviC1HDAg1sFDU6V:b57bvpx) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2b3d8450c0dcde4ec296e710fc7b1e9027cdc0f/pkg/kv/kvserver/asim/tests/testdata/non_rand/example_rebalancing.txt#L35)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Good callout, I massaged this a bit more and pushed a commit.

I need to figure out why this caused unrelated changes in the testdata output, which makes no sense to me.
</blockquote></details>

What was the reason out of curiosity? I saw it yday and didn't quite understand. 


<!-- Sent from Reviewable.io -->

wenyihu6 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> mod a comment 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153585#-:-O_MhMrV1EVa2za-FOkB:b-bxfz3n)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained

___
*[`pkg/kv/kvserver/asim/assertion/assert.go` line 144 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/153585#-O_Mfqh87XH5XJyh1P5Q:-O_Mfqh87XH5XJyh1P5R:b6mvtyo) ([raw file](https://github.com/cockroachdb/cockroach/blob/e54f3644f305410aa630866a58202b2c310ffd13/pkg/kv/kvserver/asim/assertion/assert.go#L144)):*
> ```Go
> 				// [-1, 1]. The values here would also result from the "regular"
> 				// computation below, but it doesn't hurt to be explicit.
> 				minMean = math.Inf(-1)
> ```

should we wrap this in math.Abs as well or minMean = math.Inf(1)? 


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153585#-:-O_Mi068032VDEVXCR2h:b-ydi9ay)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @wenyihu6)

___
*[`pkg/kv/kvserver/asim/assertion/assert.go` line 144 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/153585#-O_Mfqh87XH5XJyh1P5Q:-O_MhzHB7gqGF0jTAH07:bsif66k) ([raw file](https://github.com/cockroachdb/cockroach/blob/e54f3644f305410aa630866a58202b2c310ffd13/pkg/kv/kvserver/asim/assertion/assert.go#L144)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

should we wrap this in math.Abs as well or minMean = math.Inf(1)?
</blockquote></details>

:see_no_evil: thanks for catching this. Fixed.


<!-- Sent from Reviewable.io -->
